extension/src/openvic-extension/singletons/CursorSingleton.cpp

```diff
@@ -0,0 +1,622 @@
+	return rowCount*y_coord*4; //4 bytes per row * rows down
+}
+
+static int32_t _select_bits(PackedByteArray const& data, int32_t rowStart, int32_t firstBit, int32_t bitCount) {
+	int32_t bitmasks[] = {0b1 , 0b11, 0b111, 0b1111, 0b11111, 0b111111, 0b1111111, 0b11111111};
+	int32_t byteIndex = firstBit / 8;
+	int32_t bitInByteIndex = firstBit % 8;
+	if (bitInByteIndex + bitCount > 8){
+		Logger::error("Attempted to select bits outside of a byte.");
+		return 0;
+	}
+	int32_t byte = _reverse_bits(data[rowStart + byteIndex]);
+	int32_t selected = (byte >> bitInByteIndex) & bitmasks[bitCount-1];
+
+	//TODO: questionable hack, nothing in the spec suggests we should need to do this
+	if (bitCount > 1 && selected != 0){
+		return _rotate_right(selected,4);
+	}
+	return selected;
+}
+
+static bool _read_AND_mask(
+	PackedByteArray const& data, Vector2i pixelCoords, Vector2i dimensions, int32_t offset) {
+	
+	int32_t rowStart = _get_row_start(dimensions, pixelCoords.y, 1);
+	int32_t andBit = _select_bits(data, rowStart + offset,pixelCoords.x, 1);
+	return !andBit;
+}
+
+static PackedByteArray _pixel_palette_lookup(
+		PackedByteArray const& data, PackedByteArray const& palette, Vector2i coord, Vector2i dimensions,
+		int32_t offset, bool transparent, int32_t bitsPerPixel) {
+	
+	int32_t rowStart = _get_row_start(dimensions, coord.y, bitsPerPixel);
+	int32_t pixelBits = _select_bits(data, rowStart + offset, coord.x*bitsPerPixel, bitsPerPixel);
+	if ((pixelBits+1)*4 > palette.size()){
+		Logger::error("attempted to select invalid colour palette entry, ", pixelBits);
+	}
+	
+	//pixel bits serves as an index into the colour palette. We need to multiply the index by the number of bytes per colour (4)
+	PackedByteArray pixel = palette.slice(pixelBits*4,(pixelBits+1)*4);
+	pixel[3] = 0xFF * transparent;
+	return pixel;
+}
+
+/*
+24bit pixel support here is questionable:
+the spec (per daubnet) says we should pad bytes to end things on 32bit boundaries
+but the singular example of a 24bit cursor found on the internet does things like this.
+So emit a warning when trying to load one of these
+*/
+static PackedByteArray _read_24bit_pixel(
+	int32_t i, int32_t offset, PackedByteArray const& imageData, bool opaque) {
+
+	PackedByteArray pixel = PackedByteArray();
+	
+	int32_t b = imageData[offset + (i*3) + 0];
+	int32_t g = imageData[offset + (i*3) + 1];
+	int32_t r = imageData[offset + (i*3) + 2];
+
+	pixel.append(r);
+	pixel.append(g);
+	pixel.append(b);
+	pixel.append(0xFF*opaque);
+
+	return pixel;
+}
+
+static PackedByteArray _read_32bit_pixel(
+	int32_t i, int32_t offset, PackedByteArray const& imageData, bool opaque) {
+
+	PackedByteArray pixel = PackedByteArray();
+	
+	int32_t b = imageData[offset + (i*4) + 0];
+	int32_t g = imageData[offset + (i*4) + 1];
+	int32_t r = imageData[offset + (i*4) + 2];
+	int32_t a = imageData[offset + (i*4) + 3] * opaque;
+
+	pixel.append(r);
+	pixel.append(g);
+	pixel.append(b);
+	pixel.append(a);
+
+	return pixel;
+}
```

Comment:
For `_pixel_palette_lookup`, `_read_24bit_pixel` and `_read_32bit_pixel`, rather than creating a temporary `PackedByteArray` for every single pixel (each one of which requires an individual dynamic memory allocation), why not just pass the main `pixelData` array as a reference and append the bytes to it directly?
Furthermore, it would be even better if we didn't `append`/`append_array` every byte/pixel (which requires a re-allocation every time the array runs out of capacity), but instead `resize`'d the array to the final size at the beginning and then set each byte by index. You could even do this using a mutable pointer (`uint8_t*`, got from the `PackedByteArray` using `ptrw()`), so something like:
```cpp
uint8_t const* src = source_array.ptr(); // source data pointer
uint8_t* dest = destination_array.ptrw(); // destination data pointer

// This sets the byte pointed to by dest to the one pointed to by src and then moves both src and dest along to the next bytes in their repsective arrays.
*dest++ = *src++;

// If a function takes `uint8_t [const]*&` arguments then the original pointer variables given to them will also be moved along, meaning we don't have to do that manually on the caller's end.
// We don't have to move the pointers along immediately, e.g. we could do:
*dest++ = src[2];
*dest++ = src[1];
*dest++ = src[0];
*dest++ = src[3] * opaque;
src += 4;
// This is based on _read_32bit_pixel, copying a BGRA source pixel to a RGBA destination, with both pointers moved along by 4 by the end.
```

Reply:
trying to set PackedByteArray using a pointer (ie. `*arr.ptrw() = ...`) didn't seem to work for me. Just passing in a reference to the output PackedByteArray and using `[]` lets us do this just fine though.